pkg/streamingpromql/operators/function_over_range_vector.go

```diff
@@ -107,18 +118,117 @@ func (m *FunctionOverRangeVector) NextSeries(ctx context.Context) (types.Instant
 			}
 		}
 
-		accumulate(head)
-		accumulate(tail)
+		accumulate(fHead)
+		accumulate(fTail)
+
+		val := m.calculateFloatRate(step.RangeStart, step.RangeEnd, firstPoint, lastPoint, delta, fCount)
+		if data.Floats == nil {
+			// Only get fPoint slice once we are sure we have float points.
+			// This potentially over-allocates as some points in the steps may be histograms,
+			// but this is expected to be rare.
+			data.Floats, err = m.Pool.GetFPointSlice(m.numSteps)
+			if err != nil {
+				return err
+			}
+		}
+		data.Floats = append(data.Floats, promql.FPoint{T: step.StepT, F: val})
+	}
 
-		val := m.calculateRate(step.RangeStart, step.RangeEnd, firstPoint, lastPoint, delta, count)
+	if hCount >= 2 {
+		firstPoint := m.histogramBuffer.First()
+		lastPoint, _ := m.histogramBuffer.LastAtOrBefore(step.RangeEnd) // We already know there is a point at or before this time, no need to check.
 
-		data.Floats = append(data.Floats, promql.FPoint{T: step.StepT, F: val})
+		currentSchema := firstPoint.H.Schema
+		if lastPoint.H.Schema < currentSchema {
+			currentSchema = lastPoint.H.Schema
+		}
+
+		delta := lastPoint.H.CopyToSchema(currentSchema)
+		_, err = delta.Sub(firstPoint.H)
+		if err != nil {
+			return err
+		}
+		previousValue := firstPoint.H
+
+		accumulate := func(points []promql.HPoint) error {
+			for _, p := range points {
+				if p.T > step.RangeEnd { // The buffer is already guaranteed to only contain points >= rangeStart.
+					return nil
+				}
+
+				if p.H.DetectReset(previousValue) {
+					// Counter reset.
+					_, err = delta.Add(previousValue)
+					if err != nil {
+						return err
+					}
+				}
+				if p.H.Schema < currentSchema {
+					delta = delta.CopyToSchema(p.H.Schema)
+				}
+
+				previousValue = p.H
+			}
+			return nil
+		}
+
+		err = accumulate(hHead)
+		if err != nil {
+			return err
+		}
+		err = accumulate(hTail)
+		if err != nil {
+			return err
+		}
+
+		val := m.calculateHistogramRate(step.RangeStart, step.RangeEnd, firstPoint, lastPoint, delta, hCount)
+		if data.Histograms == nil {
+			// Only get hPoint slice once we are sure we have histogram points.
+			// This potentially over-allocates as some points in the steps may be floats,
+			// but this is expected to be rare.
+			data.Histograms, err = m.Pool.GetHPointSlice(m.numSteps)
+			if err != nil {
+				return err
+			}
+		}
+		val.CounterResetHint = histogram.GaugeType
+		data.Histograms = append(data.Histograms, promql.HPoint{T: step.StepT, H: val.Compact(0)})
```

Comment:
[nit] Any reason not to set the `CounterResetHint` and call `val.Compact(0)` in `calculateHistogramRate`? That feels like a better fit for its responsibilities.

Reply:
calculateHistogramRate and calculateFloatRate are very similar. I was considering deduplicating them. I don't want to split that logic out quiet yet, as we may do some more refactoring for things like `increase`. I'll put these there for now then.